clang/lib/AST/CommentParser.cpp

```diff
@@ -149,6 +149,63 @@ class TextTokenRetokenizer {
     addToken();
   }
 
+  /// Check if this line starts with @par or \par
+  bool startsWithParCommand() {
+    unsigned Offset = 1;
+
+    /// Skip all whitespace characters at the beginning.
+    /// This needs to backtrack because Pos has already advanced past the
+    /// actual \par or @par command by the time this function is called.
+    while (isWhitespace(*(Pos.BufferPtr - Offset)))
+      Offset++;
```

Comment:
You can use `StringRef::ltrim()` to handle this instead of doing it manually.

Reply:
In this case we actually have to read the buffer backwards to find the `@par`, past `Pos.BufferStart` so it's a bit inconvenient to construct the `StringRef` which starts at some unknown position before `Pos.BufferStart`. In theory someone could have put 400 spaces between `@par` and the comment text, and we'd have to accommodate that when constructing the `StringRef`. 

I think the solution as it's written right now handles that well enough. I could make it count spaces or look up the previous token containing the `@par` and compute the correct StringRef for it, but I think this is good enough.

Reply:
Okay, I think what's got me confused is that you calculate `Offset` but never actually use it (or modify `Pos.BufferPtr`). I was thinking you would have been able to use `ltrim().starts_with(...)` but I see now that the issue is we're at the start of the comment, beyond seeing `par`, and we want to backtrack.

But the logic below seems to be that we already know we're parsing a par command, so why do we need to do this check in the first place? (e.g., we get into this code path via:
```
 else if (Info->IsParCommand)
      S.actOnBlockCommandArgs(BC,
                              parseParCommandArgs(Retokenizer, Info->NumArgs));
```
and eventually land here, so don't we already know this is a par command?)

Reply:
The check is needed because Doxygen interprets the `@par` command differently if there is text on the same line or not.

If there is text on the same line, such as in Example 1 below, then the text is interpreted to be the paragraph heading, and that heading is stored as an argument for the comment. If there is no text on the same line, i.e. `@par` is alone on the line as in Example 2, then the paragraph is interpreted to have no heading.

```
/// @par I am a paragraph heading
void example1() {}

/// @par 
/// I am a paragraph with no heading
void example2() {}
```

To be able to correctly parse the paragraph heading, the lexer needs to go and check where the `@par` actually is, because if it's on the same line as some text we need to slurp that text and put it into the Arg for the comment. This code implements the functionality for this check by backtracking until it sees whitespace, and then checking whether the text immediately preceding that whitespace is a `@par` command or not.

All that being said, your feedback is helpful and I did find a small bug while re-checking this code. `Offset` should be used, otherwise the code has an implicit assumption that there is only one `char` of whitespace between the text and the `@par`. I've fixed that in the latest commit. Thank you for pointing that out :)

Reply:
Rather than backtracking, can't we still do a peek ahead? e.g., once we've seen `@par`, we can do a peek to see if the next token is a newline or not (or do a scan through the buffer to see if there is a non-whitespace character or only whitespace ending with a newline/eof)?

Reply:
The issue is that the position of `Pos.BufferPtr` is after the `@par` command when `lexParHeading` is called. I believe this is because the comment parser [consumes](https://github.com/llvm/llvm-project/blob/84dd803993fd2b6b31f8168a3f4dc729406bd3ca/clang/lib/AST/CommentParser.cpp#L407) the `@par` command token itself during parsing, and then based on the token [it decides to call](https://github.com/llvm/llvm-project/blob/84dd803993fd2b6b31f8168a3f4dc729406bd3ca/clang/lib/AST/CommentParser.cpp#L434-L436) `lexParHeading`. I've done some tests locally to verify this behavior, and my observations align with what I describe above.

While I agree that it'd be preferable to do a readahead instead of backtracking, I don't see any easy ways to refactor the PR to incorporate this behavior as it currently stands. It looks like we'd have to refactor a little more of the comment parser, which is delicate and would incorporate changes outside of the scope of this change.